File: src/test/java/com/kry/kata/bowling/BowlingGameTest.java
package com.kry.kata.bowling;

import kotlin.ranges.IntRange;
import org.junit.jupiter.api.*;

import java.util.stream.IntStream;

public class BowlingGameTest {

    private BowlingGame game;

    @BeforeEach
    void setUp() {
        game = new BowlingGame();
    }

    @Nested
    class OneTime {
        @Test
        @DisplayName("1회 투구: 1점")
        public void onl() {
            game.roll(1);

            Assertions.assertEquals(1, game.score());
        }
    }

    @Nested
    class Times {
        @Test
        @DisplayName("다회 투구: 0점")
        public void zero() {
            times(0, 20);

            Assertions.assertEquals(0, game.score());
        }

        @Test
        @DisplayName("다회 투구: 1점")
        public void one() {
            times(1, 20);

            Assertions.assertEquals(20, game.score());
        }

        @Test
        @DisplayName("다회 투구: spare")
        public void spare() {
            game.roll(3);
            game.roll(7);
            times(1, 18);

            Assertions.assertEquals(29, game.score());
        }

        @Test
        @DisplayName("다회 투구: strike")
        public void strike() {
            game.roll(10);
            times(1, 18);

            Assertions.assertEquals(30, game.score());
        }

        @Test
        @DisplayName("다회 투구: 퍼펙트")
        public void perfect() {
            times(10, 11);

            Assertions.assertEquals(300, game.score());
        }

        private void times(int pins, int times) {
            IntStream.range(0, times)
                .forEach(idx -> game.roll(pins));
        }
    }
}

File: src/test/kotlin/com/kry/kata/bowling/BowlingGameTest.kt
package com.kry.kata.bowling

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class BowlingGameTest {
    private lateinit var g: BowlingGame

    @BeforeEach
    internal fun setUp() {
        g = BowlingGame()
    }

    @Nested
    inner class `전체 스코어 점수가 같은 테스트 그룹` {
        @Test
        fun `0`() {
            20.times { g.roll(0) }

            g.assertScore(0)
        }

        @Test
        fun `1`() {
            20.times { g.roll(1) }

            g.assertScore(20)
        }
    }

    @Nested
    inner class `스페어 테스트 그룹` {
        @Test
        fun `단일 스페어 테스트`() {
            g.roll(7)
            g.roll(3)
            g.roll(3)

            17.times { g.roll(0) }

            g.assertScore(16)
        }
    }

    @Nested
    inner class `스트라이크 테스트 그룹` {
        @Test
        fun `단일 스트라이크 테스트`() {
            g.roll(10)
            g.roll(3)
            g.roll(3)

            16.times { g.roll(0) }

            g.assertScore(22)
        }
    }
}

private fun Int.times(fn: () -> Unit) = repeat(this) { fn() }
private fun BowlingGame.assertScore(expected: Int) = Assertions.assertEquals(expected, score())
